[X86] Fold BLEND(PERMUTE(X),PERMUTE(Y)) -> PERMUTE(BLEND(X,Y))

RKSimon opened the pull request:
If we don't demand the same element from both single source shuffles (permutes), then attempt to blend the sources together first and then perform a merged permute.

We don't attempt this for vXi16 blends as these are much more likely to involve byte/word vector shuffles that will result in the creation of a lot more variable-mask shuffles (PSHUFB etc.).

This fold might be worth it for VSELECT with constant masks on AVX512 targets, but I haven't investigated this yet.

The PR34592 -O0 regression is an unfortunate failure to cleanup with a later SimplifyDemandedElts pass like the -O3 does - I'm not sure how worried we should be tbh.

RKSimon committed:
[X86] Fold BLEND(PERMUTE(X),PERMUTE(Y)) -> PERMUTE(BLEND(X,Y))

If we don't demand the same element from both single source shuffles (permutes), then attempt to blend the sources together first and then perform a merged permute.

We don't attempt this for vXi16 blends as these are much more likely to involve byte/word vector shuffles that will result in the creation of a lot more variable-mask shuffles (PSHUFB etc.).

This fold might be worth it for VSELECT with constant masks on AVX512 targets, but I haven't investigated this yet.

The PR34592 -O0 regression is an unfortunate failure to cleanup with a later SimplifyDemandedElts pass like the -O3 does - I'm not sure how worried we should be tbh.

phoebewang commented on llvm/test/CodeGen/X86/horizontal-sum.ll:
```diff
@@ -679,9 +679,8 @@ define <4 x i32> @sequential_sum_v4i32_v4i32(<4 x i32> %0, <4 x i32> %1, <4 x i3
-; AVX1-SLOW-NEXT:    vpshufd {{.*#+}} xmm0 = xmm0[3,3,3,3]
-; AVX1-SLOW-NEXT:    vpblendw {{.*#+}} xmm0 = xmm0[0,1],xmm1[2,3],xmm0[4,5,6,7]
+; AVX1-SLOW-NEXT:    vpunpckhqdq {{.*#+}} xmm0 = xmm0[1],xmm1[1]
+; AVX1-SLOW-NEXT:    vpshufd {{.*#+}} xmm0 = xmm0[1,3,2,3]
```
Didn't consider posion, the two sides don't look equivalent to me.

LHS: xmm0 = xmm0[3],xmm1[3],xmm0[3],xmm0[3]
RHS: xmm0 = xmm0[3],xmm1[3],xmm1[2],xmm1[3]

I'd expect this would be `; AVX1-SLOW-NEXT: vpshufd {{.*#+}} xmm0 = xmm0[1,3,1,1]`

RKSimon replied:
The demanded elts for the original BLENDI nodes was the only the lower 2 elements - so the new permute mask ends up as <1,3,u,u> which getV4X86ShuffleImm will set the mask to use the original 2,3 elements.

I have wondered if getV4X86ShuffleImm should fold PSHUFD <X,Y,u,u> masks to <X,Y,X,Y> instead of <X,Y,2,3> but in general this doesn't help us :(

phoebewang commented:
> The PR34592 -O0 regression is an unfortunate failure to cleanup with a later SimplifyDemandedElts pass like the -O3 does - I'm not sure how worried we should be tbh.

Is it due to FastISel fall back the SelectionDAG? I don't think it's a problem here, but I'm concerning the debuggability in general if we do such optimizations under O0.

RKSimon commented:
ping?

goldsteinn commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41754,6 +41754,61 @@ bool X86TargetLowering::SimplifyDemandedVectorEltsForTargetNode(
+  case X86ISD::BLENDI: {
+    SDValue N0 = Op.getOperand(0);
+    SDValue N1 = Op.getOperand(1);
+    if (VT.getScalarSizeInBits() < 32 || !N0.hasOneUse() || !N1.hasOneUse())
```
why not for something like `blendvb`?

RKSimon replied:
Yes, I think eventually I'd like to get BLENDV/VSELECT to work with this as well. This first version is trying very hard not to introduce extra variable blend masks, hence why its just for BLENDI, but with a bit of massaging I think we can get it beneficially working for BLENDV and predicate masks as well.

goldsteinn commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41754,6 +41754,61 @@ bool X86TargetLowering::SimplifyDemandedVectorEltsForTargetNode(
+        if (Demanded0[I])
+          NewPermuteMask[I] = ScaledMask0[I];
+        else if (Demanded1[I])
+          NewPermuteMask[I] = ScaledMask1[I];
```
Need `+ NumElts` here?

RKSimon replied:
No - by blending first we've merged the 2 inputs into a single vector, so the new mask is now unary. The trick is that we never demand the same element from both sources we've blended together.

RKSimon committed:
[X86] combineBlendOfPermutes - add vXi16 (V)PBLENDW handling

v16i16 is awkward as both the new blend and permute can be higher in complexity than the original blend(permute,permute) trio.

RKSimon commented:
@goldsteinn I've pulled the combine out into its own combineBlendOfPermutes(). I've managed to get vXi16 VPBLENDW support working, with a few restrictions to avoid infinite combine loops due to creating more shuffles than we started with.....

It should be a lot easier to add BLENDV/VSELECT support in a later PR now.

RKSimon committed:
[X86] combineBlendOfPermutes - clang-format fix

goldsteinn commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -40019,6 +40029,93 @@ static SDValue combineCommutableSHUFP(SDValue N, MVT VT, const SDLoc &DL,
+  // Confirm that we only use a single operand from both permutes and that we
+  // don't demand the same index from both.
+  if (!(DemandedRHS0.isZero() && DemandedRHS1.isZero() &&
+        !DemandedLHS0.intersects(DemandedLHS1)))
```
I think this would be cleaner as: `!DemandedRHS0.isZero() || !DemandedRHS1.isZero() || DemandedLHS0.intersects(DemandedLHS1)`

goldsteinn commented:
LGTM. Wait one 1 more.

RKSimon commented:
> LGTM. Wait one 1 more.

@phoebewang  ?

phoebewang commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -40019,6 +40029,93 @@ static SDValue combineCommutableSHUFP(SDValue N, MVT VT, const SDLoc &DL,
+  assert(isBlendOrUndef(NewBlendMask) && "Bad blend");
+  assert(isUndefOrInRange(NewPermuteMask, 0, NumElts) && "Bad permute");
+
+  // v16i6 shuffles can explode in complexity very easily, only accept them if
```
v16i16?

phoebewang reviewed (approved):
LGTM.

RKSimon committed:
[X86] combineBlendOfPermutes - fix typo

KanRobert reviewed (approved):
LGTM

RKSimon committed:
[X86] combineBlendOfPermutes - simplify intersection check

dyung commented:
Hi @RKSimon, we are seeing the compiler get into a seemingly infinite loop while compiling after this change. I have put a repro in #91433. Can you take a look?
